File: Swift-C++-Testsuite/test/protocols/Swift/MethodRequirements.swift
@main
struct MyMain {

    static func main() {
        let generator = LinearCongruentialGenerator()
        print("Here's a random number: \(generator.random())")
    }

}

class LinearCongruentialGenerator: RandomNumberGenerator {
    var lastRandom = 42.0
    let m = 139968.0
    let a = 3877.0
    let c = 29573.0
    func random() -> Double {
        lastRandom =
            ((lastRandom * a + c)
                .truncatingRemainder(dividingBy: m))
        return lastRandom / m
    }
}

protocol RandomNumberGenerator {
    func random() -> Double
}

File: Swift-C++-Testsuite/test/protocols/C++/MethodRequirements.cpp

#include <iostream>
#include <math.h>


class RandomNumberGenerator
{
public:
	virtual double random() = 0;
};

class LinearCongruentialGenerator : public RandomNumberGenerator
{
private:
	double lastRandom = 42.0;
	const double m = 139968.0;
	const double a = 3877.0;
	const double c = 29573.0;

public:
	double random() {
		lastRandom = fmod((lastRandom * a + c),m);
		return lastRandom / m;
	}
};


int main()
{
		LinearCongruentialGenerator generator;
		std::cout << "Here's a random number: " << generator.random() << std::endl;
	
}